src/Components/Component.h

```diff
@@ -0,0 +1,89 @@
+#ifndef STREAM_COMPONENT_H_
+#define STREAM_COMPONENT_H_
+
+#include <thread>
+#include <atomic>
+#include <mutex>
+#include <condition_variable>
+#include <queue>
+
+namespace stream {
+	// A base class of an asynchronous FIFO task.
+	// Handles FIFO's operations and supports callbacks when task start, wait for FIFO update, process data and when task end.
+	// Example:
+	// class Encoder: public stream::Component<frame> {
+	//     void start() {
+	//         // init encoder
+	//     }
+	//     bool wait() {
+	//         // wait while queue not enpty
+	//         return getSize() == 0;
+	//	   }
+	//     void process() {
+	//	       // encoder frame and send data
+	//	   }
+	//	   void end() {
+	//	       // clear queue
+	//	       clearData();
+	//	   }
+	// }
+	// Encoder enc;
+	// // start encode
+	// enc.run();
+	// // encode frame data
+	// enc.receive(frameData);
+	// // stop encode
+	// enc.stop();
+	template<class queueType>
+	class Component {
+	public:
+		// Pushes the data into queue if it was activated.
```

Comment:
Do you have destructors and constructors? There are non-trivial member variables in this class, you'll need your destructors to clean them up properly.